### ISSUE TITLE
Allow passing kwargs to reader functions

### ISSUE THREAD

**Issue by jwindhager** [napari/napari#2278]
## 🚀 Feature

Allow passing kwargs to the reader functions, e.g. when using `napari.Viewer.open()`.

## Motivation

It took me a while to understand why my JPG was always rotated 90 degrees in napari, but not in ImageJ. Turns out that imageio's JPEG-PIL plugin, which is used by napari's built-in reader function, supports an `exifrotate` parameter (see https://imageio.readthedocs.io/en/stable/format_jpeg-pil.html#jpeg-pil). When this parameter is set to `True` (the default), imageio rotates the image depending on the value of the EXIF `Orientation` tag (which is set by various viewers, such as the Windows Photo Viewer). In most scientific imaging contexts, in my opinion, this is not the desired behavior, and the user should therefore be able to change the value of this (or other) parameters.

## Pitch

1. Add an optional `plugin_args` parameter to `napari.Viewer.open()`, which is passed to the reader function (if supported)
2. Change the reader function signature (hook specification) to optionally support `**kwargs`
3. Adapt the built-in reader function to pass the `**kwargs` to `imageio.imread()`

**Comment by tlambert03** [comment]
agreed!  this is a duplicate of #1211, (definitely considered a high value feature though!)

**Comment by jwindhager** [comment]
Apologies, didn't find this before for some reason!

**Comment by tlambert03** [comment]
> Change the reader function signature (hook specification) to optionally support **kwargs

this is the part that is potentially harder than it looks by the way.  There are restrictions in the way that hooks can be called (this is an old characteristic of pluggy that was inherited in napari-plugin-engine...)  So it may also require some surgery to the plugin engine itself to fix this.  If you're curious to dig a little bit, [`napari_plugin_engine.callers._multicall`](https://github.com/napari/napari-plugin-engine/blob/659a50f996fc35461b28073a6196b45a60e40b32/napari_plugin_engine/callers.py#L124) is basically the "core" of looping through hook implementations provided by plugins for a given hook specification.  It could go in `caller_kwargs` ... but not that the hookspec would need to have an explicit "`kwargs`" argument name, rather than `**kwargs`... and people would need to provide a dict.  (or, we'd need to do more surgery)

**Comment by sofroniewn** [comment]
Indeed, a very important topic, and one we may end up treating differently for different hookspecs.

Even just in the context of fileIO, I'll be honest, I'm still a little hesitant about opening up the hookspecs to allow arbitrary keyword arguments. If a plugin does take kwargs I'd somehow want to make sure that we can generate GUI elements for them in an open or save dialog. It would be good to collect more examples of the needs here.

In this case for example, I'd be happy to just change `exifrotate=False` and leave it, but I can see how other needs may come up and the problem is more general.

**Comment by tlambert03** [comment]
> I'm still a little hesitant about opening up the hookspecs to allow arbitrary keyword arguments.

it wouldn't be (and indeed [_can't_ be](https://github.com/napari/napari/issues/2278#issuecomment-779918184)) arbitrary **kwargs.  It would be a `plugin_kwargs` dict argument on the `open()` method (and the `read_data_with_plugins` function).  Also, it wouldn't need be on the hookspec itself ... it would be handed to the `ReaderFunction` that is _returned_ by the plugin.

```python
viewer.open(path, plugin_kwargs={'a': 1, ...})

# and then in plugins.io.read_data_with_plugins
layer_data = reader(path, **plugin_kwargs)  # try to read data
```
I think it's a pretty broad and important use case (also requested by @joshmoore in #1211).

> If a plugin does take kwargs I'd somehow want to make sure that we can generate GUI elements for them in an open or save dialog. 

I agree, that'd be nice, and could be possible if we collected some `kwargs_schema` from reader plugins.  But it seems easy to add in the future.
